### PR TITLE
Prune low-level ASN.1 parse errors from error queue in der2key_decode() etc.

### PR DESCRIPTION
**Pull request by DDvO**
The plague with many superfluous and partly misleading parse errors induced by OSSL_STORE was not over yet. 
Here is another round of cuts with the pruning machete through the jungle of wild-grown error queue insertions.

This is a follow-up to #12893 and has been motivated by (needlessly failing) tests disabled by #13022.

- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
Errors pruned for instance looked like this:
```
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_check_tlen:wrong tag::0:
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_item_embed_d2i:nested asn1 error::0:Type=X509_SIG
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_check_tlen:wrong tag::0:
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_item_embed_d2i:nested asn1 error::0:Type=EC_PRIVATEKEY
        # 40C78FACC07F0000:error::elliptic curve routines:old_ec_priv_decode:decode error::0:
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_check_tlen:wrong tag::0:
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_item_embed_d2i:nested asn1 error::0:Type=PKCS8_PRIV_KEY_INFO
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_check_tlen:wrong tag::0:
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_item_embed_d2i:nested asn1 error::0:Type=X509_PUBKEY
        # 40C78FACC07F0000:error::asn1 encoding routines:asn1_item_embed_d2i:no matching choice type::0:Type=ECPKPARAMETERS
```

**Comment by richsalz** [comment]
And what do the errors look like now?

**Comment by DDvO** [comment]
In case there is eventually a successful parse of the input file, no errors are left :smile:

And for instance when `apps/openssl pkey -in test/certs/sm2.key` is called in a configuration not supporting SM2 the output is now
```
Could not read key from test/certs/sm2.key
40A7814FA17F0000:error::elliptic curve routines:EC_GROUP_new_by_curve_name_with_libctx:unknown group:crypto/ec/ec_curve.c:3300:
Unable to load key
```

**Comment by DDvO** [comment]
Also the two Travis failures reported here are unrelated and will be fixed by #13022.

**Comment by richsalz** [comment]
So you clear the error stack when there are no errors?  That's fine.
But you also clear it when something goes wrong?  That sounds bad.

**Comment by DDvO** [comment]
> So you clear the error stack when there are no errors? That's fine.
> But you also clear it when something goes wrong? That sounds bad.

@richsalz, recently I'm not always happy with comments you give.
Some of them are simply too superficial and even misleading.

Before commenting, one should have/gain a reasonable level of understanding on the topic,
which clearly was not the case here. You should have had a look at the actual changes
and also read the discussion on the PR I have been explicitly referring to: #12893.

The error queue entries I prune here fall into two categories:
1. Errors due to soft failures of parse attempts: when OSSL_STORE loads credentials from a file
it rather blindly tries in turn a whole list of candidate formats and will succeed for at most one of them.
Such errors thrown for non-matching formats are simply garbage and should be disposed of.
2. Needless, uninformed, and partly even misleading errors thrown by some intermediate functions
after receiving a failure from a lower-level function they have called, such as these from `crypto/ec/ec_ameth.c`:
```
    eckey = eckey_type2param(ptype, pval, libctx, propq);

    if (!eckey) {
        ECerr(EC_F_ECKEY_PUB_DECODE, ERR_R_EC_LIB);
        return 0;
    }
```
and
```
    if ((ec = d2i_ECPrivateKey(NULL, pder, derlen)) == NULL) {
        ECerr(EC_F_OLD_EC_PRIV_DECODE, EC_R_DECODE_ERROR);
        return 0;
    }
```
These are particularly counterproductive when they block the pruning of errors of the first category.

**Review by t8m — APPROVED** [review]
LGTM. Please rebase to verify that after merging #13022 the (non-extented) tests still pass.

**Comment by DDvO** [comment]
> LGTM. Please rebase to verify that after merging #13022 the (non-extented) tests still pass.

Thanks @t8m for provisionally approving this.

I had found that two test cases need to be adapted
and that the output in case an EC curve is not supported should be extended to give the name of the curve.
I've also included this as a minor addendum to this PR. One of the above mentioned lines of output thus becomes:
```
C017360FBF7F0000:error::elliptic curve routines:EC_GROUP_new_by_curve_name_with_libctx:unknown group:crypto/ec/ec_curve.c:3301:name=SM2
```

File: test/recipes/25-test_x509.t
Comment (t8m) [inline test/recipes/25-test_x509.t:120]: What is the intention here? Is this just temporarily commented out?

File: test/recipes/30-test_evp.t
Comment (t8m) [inline test/recipes/30-test_evp.t:159]: Same here

File: test/recipes/25-test_x509.t
Comment (DDvO) [reply test/recipes/25-test_x509.t:120]: At first I included the `unlink $tmpfile;` since usually tmp file should be cleaned up,
but as I recently learned from @levitte it's better not to do that for the files generated in `test-runs/test_*/`
because this makes debugging tests easier and those directories are temporary anyway.
So I made a comment out of this, which was meant to stay like this.

Or should I better add a comment like `# intentionally not doing unlink $tmpfile`?

Comment (t8m) [reply test/recipes/25-test_x509.t:120]: Yeah, perhaps something like `# $tmpfile is kept to help with investigation in case of failure`

Comment (DDvO) [reply test/recipes/25-test_x509.t]: I like the suggested text; done.

File: test/recipes/30-test_evp.t
Comment (DDvO) [reply test/recipes/30-test_evp.t]: Changed as well

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Approved if the CI passes

**Comment by DDvO** [comment]
Merged - thanks @t8m

**Pull request closed without merging**
